src/goesproc/handler_dcs.cc

```diff
@@ -0,0 +1,312 @@
+  
+  s = tmpout;
+  return s;
+}
+
+std::string pseudo_decode(const char* arr,size_t len) {
+  std::string s;
+  for (unsigned int i = 0; i < len; i++)
+    s += char(arr[i] & 0x7F);
+  return s;
+}
+
+} // namespace
+
+DCSHandler::DCSHandler(
+  const Config::Handler& config,
+  const std::shared_ptr<FileWriter>& fileWriter)
+  : config_(config),
+    fileWriter_(fileWriter) {
+}
+
+void DCSHandler::getHeader(char *const buf) {
+  size_t nread = 0;
+  // 32 bytes hold the DCS file name (and trailing spaces)
+  {
+    constexpr unsigned n = 32;
+    header.filename = std::string(buf + nread, n);
+    nread += n;
+  }
+
+  // 8 bytes holding length of payload
+  {
+    constexpr unsigned n = 8;
+    header.file_size = std::stoi(std::string(buf + nread, n));
+    nread += n;
+  }
+
+  // 4 bytes holding the file's source (WCDA or NSOF)
+  {
+    constexpr unsigned n = 4;
+    header.file_source = std::string(buf + nread, n);
+    nread += n;
+  }
+
+  // 4 bytes holding the file type, should be DCSH (DCS HRIT)
+  {
+    constexpr unsigned n = 4;
+    header.file_type = std::string(buf + nread, n);
+    nread += n;
+  }
+
+  // 12 bytes holding spaces, reserved for future usage, should be all spaces
+  {
+    constexpr unsigned n = 12;
+    header.exp_fill = std::string(buf + nread, n);
+    nread += n;
+  }
+
+  // 4 bytes holding the CRC32 of the header, little-endian
+  {
+    constexpr unsigned n = 4;
+    
+    nread += n;
+  }
+
+}
+
+void DCSHandler::handle(std::shared_ptr<const lrit::File> f) {
+  auto ph = f->getHeader<lrit::PrimaryHeader>();
+  if (ph.fileType != 130) {
+    return;
+  }
+
+  auto text = f->getHeader<lrit::AnnotationHeader>().text;
+  struct timespec time = {0, 0};
+
+  if (!goesrParseDCSTime(text, time)) {
+      // Unable to extract timestamp from file name
+  }
+
+
+  // Skip if the time could not be determined
+  if (time.tv_sec == 0) {
+    return;
+  }
+
+  // Pointer to beginning of the DCS file's data
+  char *const buf = (f->read()).data();
+  
+  // Parse the DCS file's header
+  DCSHandler::getHeader(buf);
+
+  // Create the output data by sending text into a stream
+  std::stringstream output_text;
+  output_text << "[DCS Header]" << std::endl;
+  output_text << "File Name: " << text << std::endl;
+  output_text << "File Size: " << std::dec << header.file_size << " Bytes" << std::endl;
+  output_text << "File Source: " << header.file_source << std::endl;
+  output_text << "File Type: " << header.file_type << std::endl;
+  output_text << "Header CRC32: " << std::hex << header.crc32 << std::endl << std::endl;
+
+  size_t offset = 64;
+  size_t nread = 0;
+  while(offset < header.file_size-4) {
+    nread = 0;
+    DCPBlock block;
+
+    { // Block Type
+      constexpr unsigned n = 1;
+      block.type = bytes_to_uint(buf + offset + nread, n);
+      nread += n;
+    }
+    { // Block Length
+      constexpr unsigned n = 2;
+      block.size = bytes_to_uint(buf + offset + nread, n);
```

Comment:
Looks like the issue could be with this function, ends up reading the block size incorrectly in some cases, could be the reason the whole file doesn't get parsed either, still not sure myself why it would be having issues though.
https://github.com/pietern/goestools/blob/40e2a62aaa52fda91b5fe09a138a4e1c88251cf0/src/goesproc/handler_dcs.cc#L39-L45

Reply:
Looks like this fixed the issue:
https://github.com/pietern/goestools/blob/baf6958d2a12174f2d81eff261301984ba331ae8/src/goesproc/handler_dcs.cc#L39-L45